### PR TITLE
Pr/pdm

### PR DESCRIPTION
**Pull request by mwever**
Enable ML-Plan using scikit-learn and tsfresh to deal with predictive maintenance tasks.

### CODE REVIEW

File: softwareconfiguration/mlplan/sklearn/conf/log4j.dtd [hunk 1]
```diff
@@ -0,0 +1,227 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (alexandertornede) [inline softwareconfiguration/mlplan/sklearn/conf/log4j.dtd:1]: Are we sure that we want the logging files in these commits?

File: softwareconfiguration/mlplan/sklearn/resources/automl/searchmodels/sklearn/sklearn-preprocessors.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (alexandertornede) [inline softwareconfiguration/mlplan/sklearn/resources/automl/searchmodels/sklearn/sklearn-preprocessors.json:1]: Why is there only a single preprocessor contained?

File: softwareconfiguration/mlplan/sklearn/resources/automl/searchmodels/sklearn/sklearn-rul.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+				{
+					"id": "timeseries_transformer",
+					"name": "TimeseriesTransformer"
+				},
+				{
+					"id": "data_cleaner",
+					"name": "DataCleaner"
+				},
+				{
+					"id": "preprocessor0",
```
Comment (alexandertornede) [inline softwareconfiguration/mlplan/sklearn/resources/automl/searchmodels/sklearn/sklearn-rul.json:26]: Why do we have more than 1 preprocessor?

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/core/evaluation/evaluator/factory/MonteCarloCrossValidationEvaluatorFactory.java [hunk 4]
```diff
@@ -29,7 +29,7 @@ public MonteCarloCrossValidationEvaluator getLearnerEvaluator() {
 			throw new IllegalStateException("Cannot create MCCV evaluator due to invalid number of repeats " + this.getNumMCIterations() + ". Set number of repeats to a positive value!");
 		}
 		IAggregatedPredictionPerformanceMeasure<?, ?> aggMeasure = new AveragingPredictionPerformanceMeasure<>(this.metric);
-		return new MonteCarloCrossValidationEvaluator(this.data, this.getNumMCIterations(), this.getTrainFoldSize(), this.random, aggMeasure);
+		return new MonteCarloCrossValidationEvaluator(this.getCacheSplitSets(), this.data, this.getNumMCIterations(), this.getTrainFoldSize(), this.random, aggMeasure);
```
[the remark this answers is not in the document]
Comment (mwever) [reply JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/core/evaluation/evaluator/factory/MonteCarloCrossValidationEvaluatorFactory.java:32]: yes

File: softwareconfiguration/mlplan/core/src/main/java/ai/libs/mlplan/safeguard/EvaluationSafeGuardFiredEvent.java [hunk 5]
```diff
@@ -20,7 +20,7 @@ public ComponentInstance getComponentInstance() {
 
 	@Override
 	public long getTimestamp() {
-		return 0;
```
[the remark this answers is not in the document]
Comment (mwever) [reply softwareconfiguration/mlplan/core/src/main/java/ai/libs/mlplan/safeguard/EvaluationSafeGuardFiredEvent.java:23]: Code smell

File: softwareconfiguration/mlplan/sklearn/conf/log4j.dtd (hunk 1 above)
Comment (mwever) [reply softwareconfiguration/mlplan/sklearn/conf/log4j.dtd:1]: why not?

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/core/evaluation/evaluator/TrainPredictionBasedClassifierEvaluator.java [hunk 6]
```diff
@@ -92,17 +93,18 @@ public Double evaluate(final ISupervisedLearner<ILabeledInstance, ILabeledDatase
 						}
 					}
 					if (m - mistakes == 0) {
-						this.logger.warn("0 correct predictions seems suspicious. Here are the vectors: \n\tGround truth: {}\n\tPredictions: {}", report.getPredictionDiffList().getGroundTruthAsList(),
+						this.logger.debug("0 correct predictions seems suspicious. Here are the vectors: \n\tGround truth: {}\n\tPredictions: {}", report.getPredictionDiffList().getGroundTruthAsList(),
```
[the remark this answers is not in the document]
Comment (tornede) [reply JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/core/evaluation/evaluator/TrainPredictionBasedClassifierEvaluator.java:96]: Done.

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/core/evaluation/evaluator/TrainPredictionBasedClassifierEvaluator.java [hunk 7]
```diff
@@ -54,12 +55,12 @@ public Double evaluate(final ISupervisedLearner<ILabeledInstance, ILabeledDatase
...
 			for (int i = 0; i < n; i++) {
 				List<ILabeledDataset<? extends ILabeledInstance>> folds = splitSet.getFolds(i);
-				this.logger.debug("Executing learner{} on folds of sizes {} (train) and {} (test) using {}.", learner, folds.get(0).size(), folds.get(1).size(), this.executor.getClass().getName());
+				this.logger.debug("Executing learner {} on folds of sizes {} (train) and {} (test) using {}.", learner, folds.get(0).size(), folds.get(1).size(), this.executor.getClass().getName());
 				ILearnerRunReport report;
 				try {
 					report = this.executor.execute(learner, folds.get(0), folds.get(1));
-					this.logger.trace("Obtained report. Training times was {}ms, testing time {}ms. Ground truth vector: {}, prediction vector: {}", report.getTrainEndTime() - report.getTrainStartTime(),
-							report.getTestEndTime() - report.getTestStartTime(), report.getPredictionDiffList().getGroundTruthAsList(), report.getPredictionDiffList().getPredictionsAsList());
+					this.logger.info("Obtained report. Training times was {}ms, testing time {}ms. Ground truth vector: {}, prediction vector: {}. Pipeline: {}", report.getTrainEndTime() - report.getTrainStartTime(),
```
[the remark this answers is not in the document]
Comment (tornede) [reply JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/core/evaluation/evaluator/TrainPredictionBasedClassifierEvaluator.java:62]: Done.

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/regression/loss/dataset/AsymmetricLoss.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
+package ai.libs.jaicore.ml.regression.loss.dataset;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import ai.libs.jaicore.basic.StatisticsUtil;
+
+public class AsymmetricLoss extends ARegressionMeasure {
+
+	private double dividerUnderestimation = 5;
```
[the remark this answers is not in the document]
Comment (tornede) [reply JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/regression/loss/dataset/AsymmetricLoss.java:10]: Done.

File: JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/regression/singlelabel/SingleTargetRegressionPredictionBatch.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+public class SingleTargetRegressionPredictionBatch extends ArrayList<IRegressionPrediction> implements IRegressionResultBatch {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(SingleTargetRegressionPredictionBatch.class);
+
+	private static final long serialVersionUID = 1L;
+
+	public SingleTargetRegressionPredictionBatch(final Collection<IRegressionPrediction> predictions) {
+		this.addAll(predictions);
+		LOGGER.debug("Predictions: {}", predictions.stream().map(IRegressionPrediction::getDoublePrediction).collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (tornede) [reply JAICore/jaicore-ml/src/main/java/ai/libs/jaicore/ml/regression/singlelabel/SingleTargetRegressionPredictionBatch.java:21]: Actually, this was only added for debugging. Removed the logger output.

File: softwareconfiguration/mlplan/sklearn/resources/automl/searchmodels/sklearn/sklearn-rul.json (hunk 3 above)
Comment (tornede) [reply softwareconfiguration/mlplan/sklearn/resources/automl/searchmodels/sklearn/sklearn-rul.json:26]: Replaced search space to be up to date.

File: softwareconfiguration/mlplan/sklearn/resources/automl/searchmodels/sklearn/sklearn-preprocessors.json (hunk 2 above)
Comment (tornede) [reply softwareconfiguration/mlplan/sklearn/resources/automl/searchmodels/sklearn/sklearn-preprocessors.json:1]: Replaced search space to be up to date.
